### DISCUSSION TITLE
Dictionaries vs. enums for defining instrument controls

### DISCUSSION THREAD

**Discussion by CodingMarco** [pymeasure/pymeasure#560] [General]
I see that all currently implemented instrument drivers (apart from AgilentB1500) use dictionaries to define instrument controls (modes, waveform shapes etc.).
Is there a specific reason for that? In my opinion, enums would be much cleaner than dictionaries.
Many editors/IDEs have an autocompletion feature, that, combined with using enums, would make it much easier to get started with using a driver that you don't already know.

Of course it would be best to use the same in all drivers, but am I "allowed" to use enums instead of dictionaries in my drivers or is this discouraged?

**Comment by msmttchr** [comment]
I think there's no guideline to avoid enums.
Normally dictionaries are used to map values, but if you have good reason to use an enum instead, why not?
Please note that, on the other hand, property factories like `Instrument.(control|measure|setting)` accept only  list, tuple, range, or dictionary as `values` parameter.

**Reply by bilderbuchi** [reply]
Yeah, dictionaries are used to define a mapping between user-facing, easily understandable values and what the instruments expects. They are also inverted internally to cater for both directions of mapping. 
I think in most cases you could do this with enums, too, in principle, but the factories cannot deal with this.
The current best practice is to have a good docstring.
